README.md
Tapalcatl Cookbook
=================

This cookbook sets up [Tapalcatl](https://github.com/tilezen/tapalcatl), the "metatile" server.

Requirements
------------

#### recipes
- `apt`
- `runit`
- `user`

Attributes
----------

#### tapalcatl::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['package']</tt></td>
    <td>String</td>
    <td>Package to install</td>
    <td><tt>github.com/tilezen/tapalcatl/tapalcatl_server</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['cfg_file']</tt></td>
    <td>String</td>
    <td>Config file name</td>
    <td><tt>tapalcatl.conf</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['cfg_path']</tt></td>
    <td>String</td>
    <td>Config file location</td>
    <td><tt>/etc/tapalcatl</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['pattern']</tt></td>
    <td>JSON object</td>
    <td>
        <p>A map of Gorilla Mux patterns to objects describing the metatile storage and upstream server fallback. Each pattern maps a distinct URL path on the Tapalcatl server to a storage source, layer, upstream proxy URL and metatile size. The storage source can be either 'File' for local file storage (which takes a 'BaseDir' subkey) or 'S3' to use S3 (which takes 'Bucket', 'KeyPattern' and 'Prefix' subkeys).</p>
        <p>The S3 KeyPattern and upstream ProxyURL are both interpolated with the variables matches against the input pattern, plus the Layer parameter. The S3 KeyPattern additionally can contain {hash} which is a partial md5 of the pattern <tt>'{layer}/{z}/{x}/{y}.{fmt}'</tt> useful for distributing the filenames.</p>
    </td>
    <td>It's complicated. See the default attributes file for examples.</td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['listen']</tt></td>
    <td>String</td>
    <td>The `interface:port` to listen on.</td>
    <td><tt>:80</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['healthcheck']</tt></td>
    <td>String</td>
    <td>Optional. If present, then the path to respond to healthchecks on.</td>
    <td><tt>nil</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['mime']</tt></td>
    <td>JSON object</td>
    <td>A map of request extensions to MIME types. This is used to lookup a MIME type for the response headers.</td>
    <td>See the default attributes file for an example.</td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['poolsize']</tt></td>
    <td>Integer</td>
    <td>Tapalcatl has a buffer pool to reduce GC pressure when copying large amounts of data from the upstream server. The poolsize is the number of buffers to retain, and should be sized to the number of concurrent requests expected.</td>
    <td>10</td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['poolwidth']</tt></td>
    <td>Integer</td>
    <td>The pool width is the size in bytes to initially assign. Buffers can grow, so it is not necessary to make this particularly large. A good value is the average expected size of the zip files.</td>
    <td>102400</td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['runit']['timeout']</tt></td>
    <td>Integer</td>
    <td>The time in seconds runit should wait to see if a service starts up successfully.</td>
    <td><tt>60</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['user']['create_group']</tt></td>
    <td>Boolean</td>
    <td>Whether to create a group for the user.</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['user']['enabled']</tt></td>
    <td>Boolean</td>
    <td>Whether to create a user for tapalcatl.</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['user']['home']</tt></td>
    <td>String</td>
    <td>The user's home directory location.</td>
    <td><tt>/home/tapalcatl</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['user']['user']</tt></td>
    <td>String</td>
    <td>The user's name.</td>
    <td><tt>tapalcatl</tt></td>
  </tr>
  <tr>
    <td><tt>['tapalcatl']['limits']['open_files']</tt></td>
    <td>Integer</td>
    <td>The maximum number of open file descriptors allowed.</td>
    <td>The value in <tt>/proc/sys/fs/file-max</tt></td>
  </tr>
</table>

Contributing
------------

We welcome contributions to tapalcatl's Chef recipe. To contribute, please use the standard Github workflow:

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------

Copyright Mapzen 2016. Available under the MIT license. See LICENSE for more information.


CHANGELOG.md
tapalcatl CHANGELOG
===================

0.3.1
-----

* Increase limit on open file descriptors. [#7](https://github.com/tilezen/chef-tapalcatl/pull/7).

0.3.0
-----

* Simplify the default tapalcatl configuration.

0.2.0
-----

* Update configuration
  - Support new tapalcatl storage configuration
  - Support expvars configuration
* Do not prefix timestamps from runit

0.1.1
-----

* Update generation of url used to pull down binary.

0.1.0
-----

* Initial


recipes/install.rb
#
# Cookbook Name:: tapalcatl
# Recipe:: install
#
# Copyright 2016, Mapzen
#
# Available under the MIT license, see LICENSE for details
#

include_recipe "runit"

template "#{node[:tapalcatl][:cfg_path]}/#{node[:tapalcatl][:cfg_file]}" do
  source 'tapalcatl-config.conf.erb'
  notifies :restart, 'runit_service[tapalcatl]', :delayed
end

url = "https://s3.amazonaws.com/mapzen.software/tile/tapalcatl/tapalcatl_server-#{node[:tapalcatl][:revision]}"
remote_file "#{node[:tapalcatl][:bin]}/tapalcatl_server" do
  source url
  mode '0755'
  action :create
  notifies :restart, 'runit_service[tapalcatl]', :delayed
end

execute 'allow tapalcatl to bind on <1024 ports' do
  command 'setcap cap_net_bind_service=+ep ' + node[:tapalcatl][:bin] + '/tapalcatl_server'
end

# create directory, since runit doesn't seem to do this automatically
directory '/var/log/tapalcatl' do
  action :create
  recursive true
end

# set the default to the max if there's no default set
ruby_block "set default ulimit" do
  block do
    unless node.default[:tapalcatl][:limits][:open_files]
      max = File.read('/proc/sys/fs/file-max').to_i
      # sanity check
      if max >= 1024
        node.default[:tapalcatl][:limits][:open_files] = max
      end
    end
  end
end

runit_service 'tapalcatl' do
  action [:enable]
  log true
  default_logger false
  sv_timeout node[:tapalcatl][:runit][:timeout]
end


attributes/default.rb
default[:tapalcatl][:bin] = '/usr/local/bin'
default[:tapalcatl][:cfg_file] = 'tapalcatl.conf'
default[:tapalcatl][:cfg_path] = '/etc/tapalcatl'
default[:tapalcatl][:revision] = 'master'

# interface:port to listen on
default[:tapalcatl][:listen] = ':80'

# each pattern maps a distinct URL path on the tapalcatl server to a storage
# source, layer, upstream proxy URL and metatile size. the storage source can
# be either 'File' for local file storage (which takes a 'BaseDir' subkey) or
# 'S3' to use S3 (which takes 'Bucket', 'KeyPattern' and 'Prefix' subkeys).
#
# the S3 KeyPattern and upstream ProxyURL are both interpolated with the
# variables matches against the input pattern, plus the Layer parameter. the
# S3 KeyPattern additionally can contain {hash} which is a partial md5 of the
# pattern '{layer}/{z}/{x}/{y}.{fmt}' useful for distributing the filenames.
#

# a mapping from the request extension to the MIME type to use in the returned
# headers.
default[:tapalcatl][:mime] = {
  'json' => 'application/json',
  'mvt' => 'application/x-protobuf',
  'mvtb' => 'application/x-protobuf',
  'topojson' => 'application/json',
}

# expvar configuration
default[:tapalcatl][:expvars][:serve] = false
default[:tapalcatl][:expvars][:log_interval] = 0

handler_cfg = {
  'mime' => node[:tapalcatl][:mime],
  'storage' => {
    'tmp' => {
      'type' => 'file',
      'basedir' => '/tmp',
      'metatilesize' => 1,
      'layer' => 'all',
    },
  },
  'pattern' => {
    '/{z:[0-9]+}/{x:[0-9]+}/{y:[0-9]+}.{fmt}' => {
      'storage' => 'tmp',
    },
  },
}
default[:tapalcatl][:handler_cfg_json] = handler_cfg.to_json

# tapalcatl has a buffer pool to reduce GC pressure when copying large amounts
# of data from the upstream server.
default[:tapalcatl][:pool][:enabled] = true

# the "num" is the number of buffers to retain, and should be sized to
# the number of concurrent requests expected.
default[:tapalcatl][:pool][:num] = 10

# the pool entry is the size in bytes to initially assign. buffers can grow, so
# it is not necessary to make this particularly large. a good value is the
# average expected size of the zip files.
default[:tapalcatl][:pool][:entry] = 102400

default[:tapalcatl][:runit][:timeout] = 60
default[:tapalcatl][:user][:create_group] = true
default[:tapalcatl][:user][:enabled] = true
default[:tapalcatl][:user][:home] = '/home/tapalcatl'
default[:tapalcatl][:user][:user] = 'tapalcatl'

default[:tapalcatl][:metrics][:enabled] = false
# addr should be a "host:port"
default[:tapalcatl][:metrics][:statsd][:addr] = ''
# prefix used when sending metrics to statsd
default[:tapalcatl][:metrics][:statsd][:prefix] = ''

# default to using whatever the maximum is that the system advertises in
# /proc/sys/fs/file-max. override to set a specific (lower) value.
default[:tapalcatl][:limits][:open_files] = nil
